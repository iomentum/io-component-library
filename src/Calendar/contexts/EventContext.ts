import { createContext, Dispatch, SetStateAction } from 'react';
import { EventAction, Event } from '../reducers/EventReducer';

export interface EventContextInterface {
  eventsCollection: Event[];
  setEventsCollection: Dispatch<SetStateAction<Event[]>>;
  eventManagementOpened: boolean;
  setEventManagementOpened: Dispatch<SetStateAction<boolean>>;
  event: Event;
  dispatchEvent: Dispatch<EventAction>;
}

export const EventContext = createContext<EventContextInterface>(null);
