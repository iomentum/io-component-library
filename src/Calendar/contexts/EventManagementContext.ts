import { createContext, Dispatch, SetStateAction } from 'react';
import { EventAction, EventModel } from '../reducers/EventReducer';

export interface EventManagementContext {
  fullDayEvent: boolean;
  event: EventModel;
  setFullDayEvent: Dispatch<SetStateAction<boolean>>;
  dispatchEvent: Dispatch<EventAction>;
}

export const EventManagementContext = createContext<EventManagementContext>(null);
