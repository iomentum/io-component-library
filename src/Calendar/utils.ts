import * as m from 'moment';
import { DateRange, extendMoment, MomentRangeStaticMethods } from 'moment-range';
import Dayz from 'dayz';

export const extendedMoment = extendMoment(m);

export type MomentRangeExtended = MomentRangeStaticMethods & m.Moment;

export enum DisplayMode {
  Week = 'week',
  Day = 'day',
}

export interface Event {
  content: string;
  range: () => DateRange;
}

export interface EventsCollection {
  events: Event[];
  add: (
    eventAttrs: {
      content: string;
      range: DateRange;
    },
    options?: Record<string, never>
  ) => void;
}

export const hours = [
  '00:00',
  '00:15',
  '00:30',
  '00:45',
  '01:00',
  '01:15',
  '01:30',
  '01:45',
  '02:00',
  '02:15',
  '02:30',
  '02:45',
  '03:00',
  '03:15',
  '03:30',
  '03:45',
  '04:00',
  '04:15',
  '04:30',
  '04:45',
  '05:00',
  '05:15',
  '05:30',
  '05:45',
  '06:00',
  '06:15',
  '06:30',
  '06:45',
  '07:00',
  '07:15',
  '07:30',
  '07:45',
  '08:00',
  '08:15',
  '08:30',
  '08:45',
  '09:00',
  '09:15',
  '09:30',
  '09:45',
  '10:00',
  '10:15',
  '10:30',
  '10:45',
  '11:00',
  '11:15',
  '11:30',
  '11:45',
  '12:00',
  '12:15',
  '12:30',
  '12:45',
  '13:00',
  '13:15',
  '13:30',
  '13:45',
  '14:00',
  '14:15',
  '14:30',
  '14:45',
  '15:00',
  '15:15',
  '15:30',
  '15:45',
  '16:00',
  '16:15',
  '16:30',
  '16:45',
  '17:00',
  '17:15',
  '17:30',
  '17:45',
  '18:00',
  '18:15',
  '18:30',
  '18:45',
  '19:00',
  '19:15',
  '19:30',
  '19:45',
  '20:00',
  '20:15',
  '20:30',
  '20:45',
  '21:00',
  '21:15',
  '21:30',
  '21:45',
  '22:00',
  '22:15',
  '22:30',
  '22:45',
  '23:00',
  '23:15',
  '23:30',
  '23:45',
];

export const fromDateForHours = (date: Date) => {
  const min = date.getMinutes();

  if (min < 15) {
    return hours[date.getHours() * 4];
  }
  if (min < 30) {
    return hours[date.getHours() * 4 + 1];
  }
  if (min < 45) {
    return hours[date.getHours() * 4 + 2];
  }
  return hours[date.getHours() * 4 + 3];
};

export const getDefaultEvent = (start: MomentRangeExtended): Event => ({
  content: '',
  range: () => extendedMoment.range(extendedMoment(start), extendedMoment(start).add(1, 'hour')),
});

interface SelectedEvent {
  content: string;
  range: DateRange;
}

export const findEvent = (event: SelectedEvent, events: EventsCollection) =>
  events.events.find(
    (currEvent) => event.content === currEvent.content && event.range.isSame(currEvent.range())
  );

// const date = extendedMoment();

export const EVENTS: EventsCollection = new Dayz.EventsCollection([
  // {
  //   content: "Weeklong",
  //   range: extendedMoment.range(date.clone(), date.clone().endOf("day")),
  // },
  // {
  //   content: "9am - 2pm",
  //   range: extendedMoment.range(date.clone().hour(9), date.clone().hour(14)),
  // },
  // {
  //   content: "8am - 8pm",
  //   range: extendedMoment.range(date.clone().hour(8), date.clone().hour(21).minutes(40)),
  // },
]);
