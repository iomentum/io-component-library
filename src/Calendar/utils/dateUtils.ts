export const getFormattedDate = (date: Date) => date.toISOString().replace(/T.*/, '');

export const formatDateAndHour = (dateToFormat: Date): [string, string] => {
  const timeZoneOffset = dateToFormat.getTimezoneOffset() * 60000;
  const [date, hour] = new Date(dateToFormat.getTime() - timeZoneOffset).toISOString().split('T');
  return [date, hour.slice(0, 5)];
};

export const mergeDateAndHour = (date: Date, hour: string) => {
  const [hours, minutes] = hour.split(':');
  return new Date(date.setHours(+hours, +minutes));
};
