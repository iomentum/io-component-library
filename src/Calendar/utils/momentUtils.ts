import moment, * as m from 'moment';
import { MomentInput } from 'moment';
import { extendMoment } from 'moment-range';

const extendedMoment = extendMoment(m);

export const createExtendedMomentFromDate = (date: Date) => extendedMoment(date);

export const createMomentRange = (startDate: Date, endDate: Date) =>
  extendedMoment.range(extendedMoment(new Date(startDate)), extendedMoment(new Date(endDate)));

export const createDateFromMoment = (momentInput: MomentInput) => moment(momentInput).toDate();
