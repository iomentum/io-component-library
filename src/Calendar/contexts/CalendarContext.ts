import { createContext, Dispatch, SetStateAction } from "react";
import { Display, MomentRangeExtended } from "../utils";

export interface CalendarContextInterface {
  display: Display;
  setDisplay: Dispatch<SetStateAction<Display>>;
  setDate: Dispatch<SetStateAction<MomentRangeExtended>>;
}

export const CalendarContext = createContext<CalendarContextInterface>(null);
