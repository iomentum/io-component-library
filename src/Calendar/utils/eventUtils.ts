import { DateRange } from 'moment-range';
import Dayz from 'dayz';
import { EventModel } from '../reducers/EventReducer';
import { Event, extendedMoment } from '../utils';

export interface MyCalendarEvent {
  uuid: string;
  title: string;
  startDate: Date;
  endDate: Date;
  metadata: Record<string, unknown>;
}

export interface DayzEvent {
  content: string;
  range: DateRange;
  uuid: string;
}

export interface DayzEventsCollection {
  events: DayzEvent[];
  add: (
    eventAttrs: {
      content: string;
      range: DateRange;
    },
    options?: Record<string, never>
  ) => void;
}

export const currentEventIndex = (
  myCalendarEvents: MyCalendarEvent[],
  myCalendarEvent: MyCalendarEvent
) => myCalendarEvents.findIndex((event) => event.uuid === myCalendarEvent.uuid);

export const deleteEvent = (
  myCalendarEvents: MyCalendarEvent[],
  myCalendarEvent: MyCalendarEvent
) => {
  const eventToDeleteIndex = currentEventIndex(myCalendarEvents, myCalendarEvent);

  myCalendarEvents.splice(eventToDeleteIndex, 1);
};

export const isEventExisting = (
  eventsCollection: MyCalendarEvent[],
  myCalendarEvent: MyCalendarEvent
) => currentEventIndex(eventsCollection, myCalendarEvent) !== -1;

const computeEventDate = (event: EventModel): [Date, Date] => {
  const [newEventStartHour, newEventStartMinutes] = event.startHour.split(':');
  const [newEventEndHour, newEventEndMinutes] = event.endHour.split(':');

  const startDate = new Date(event.startDate);
  startDate.setHours(+newEventStartHour, +newEventStartMinutes);

  const endDate = new Date(event.endDate);
  endDate.setHours(+newEventEndHour, +newEventEndMinutes);

  return [startDate, endDate];
};

export const createEvent = (event: EventModel): MyCalendarEvent => {
  const [startDate, endDate] = computeEventDate(event);

  return {
    uuid: 'awdawd',
    title: event.content,
    startDate,
    endDate,
    metadata: {},
  };
};

export const updateEvent = (event: EventModel, eventToUpdate: MyCalendarEvent): MyCalendarEvent => {
  const [startDate, endDate] = computeEventDate(event);

  return {
    uuid: eventToUpdate.uuid,
    title: event.content,
    startDate,
    endDate,
    metadata: eventToUpdate.metadata,
  };
};

const convertMyCalendarEventIntoDayzEvent = (myCalendarEvent: MyCalendarEvent): DayzEvent => ({
  content: myCalendarEvent.title,
  range: extendedMoment.range(
    extendedMoment(new Date(myCalendarEvent.startDate)),
    extendedMoment(new Date(myCalendarEvent.endDate))
  ),
  uuid: myCalendarEvent.uuid,
});

export const convertMyCalendarEventsIntoDayzEventsCollection = (
  myCalendarEvents: MyCalendarEvent[]
): DayzEventsCollection =>
  new Dayz.EventsCollection([...myCalendarEvents.map(convertMyCalendarEventIntoDayzEvent)]);

export const createDefaultMyCalendarEvent = (startDate: Date): MyCalendarEvent => {
  const endDate = new Date(startDate);
  endDate.setHours(endDate.getHours() + 1);

  return {
    uuid: `awdawd`,
    title: '',
    startDate,
    endDate,
    metadata: {},
  };
};

export const findEvent = (event: DayzEvent, myCalendarEvents: MyCalendarEvent[]) =>
  myCalendarEvents.find((myCalendarEvent) => myCalendarEvent.uuid === event.uuid);
