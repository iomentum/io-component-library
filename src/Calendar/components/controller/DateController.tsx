import React, {
  memo,
  MouseEventHandler,
  useCallback,
  useContext,
  useRef,
} from "react";
import { Button } from "@material-ui/core";
import { CalendarContext } from "../../contexts/CalendarContext";
import { extendedMoment } from "../../utils";

interface ControlButtonProps {
  label: string;
  onClick: MouseEventHandler<HTMLButtonElement>;
}

export const ControlButton = (props: ControlButtonProps) => (
  <Button onClick={props.onClick}>{props.label}</Button>
);

export const DateController = () => {
  const { display, setDate } = useContext(CalendarContext);

  const handleSubstractButton = useCallback(
    () => setDate((date) => extendedMoment(date.subtract(1, display))),
    [setDate, display]
  );

  const handleTodayButton = useCallback(() => setDate(() => extendedMoment()), [
    setDate,
  ]);

  const handleAddButton = useCallback(
    () => setDate((date) => extendedMoment(date.add(1, display))),
    [setDate, display]
  );

  return (
    <>
      <ControlButton onClick={handleSubstractButton} label={"<"} />
      <ControlButton onClick={handleTodayButton} label={"Today"} />
      <ControlButton onClick={handleAddButton} label={">"} />
    </>
  );
};
