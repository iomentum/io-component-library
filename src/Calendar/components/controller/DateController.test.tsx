/**
 * @jest-environment jsdom
 */

import { render, screen } from "@testing-library/react";
import React from "react";
import { CalendarContext } from "../../contexts/CalendarContext";
import { Display } from "../../utils";
import { ControlButton, DateController } from "./DateController";

const calendarContextMock = (component, { providerProps, ...renderOptions }) =>
  render(
    <CalendarContext.Provider value={providerProps}>
      {component}
    </CalendarContext.Provider>,
    renderOptions
  );

describe("ControlButton component", () => {
  describe("@snapshot", () => {
    it("should match with previous ControlButton", () => {
      const { asFragment } = render(
        <ControlButton label="test" onClick={() => {}} />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("@event", () => {
    it("should handle a click", () => {
      const handleOnClick = jest.fn();
      render(<ControlButton label="" onClick={handleOnClick} />);

      screen.getByRole("button").click();

      expect(handleOnClick.mock.calls.length).toEqual(1);
    });
  });
});

describe("DateController component", () => {
  describe("@snapshot", () => {
    it("should match with previous DateController", () => {
      const providerProps = {
        display: Display,
        setDate: jest.fn(),
      };
      const { asFragment } = calendarContextMock(<DateController />, {
        providerProps,
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("@event", () => {
    it("each buttons should trigger a setDate on click", () => {
      const providerProps = {
        display: Display,
        setDate: jest.fn(),
      };
      calendarContextMock(<DateController />, { providerProps });

      screen.getAllByRole("button").forEach((button) => button.click());

      expect(providerProps.setDate.mock.calls.length).toEqual(3);
    });
  });
});
