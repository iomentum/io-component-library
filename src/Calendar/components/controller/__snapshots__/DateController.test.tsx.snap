// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlButton component @snapshot should match with previous ControlButton 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      test
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`DateController component @snapshot should match with previous DateController 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      &lt;
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Today
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      &gt;
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
